src/vcpkg/binarycaching.cpp

```diff
@@ -1419,21 +1430,30 @@ namespace
             return msg::format(msgRestoredPackagesFromAZUPKG, msg::count = count, msg::elapsed = ElapsedTime(elapsed));
         }
 
-        void fetch(View<const InstallPlanAction*> actions, Span<RestoreResult> out_status) const override
+        void acquire_zips(View<const InstallPlanAction*> actions, Span<Optional<ZipResource>> out_zips) const override
         {
             for (size_t i = 0; i < actions.size(); ++i)
             {
-                auto info = BinaryPackageReadInfo{*actions[i]};
-                auto ref = make_feedref(info, "");
-                auto res = m_azure_tool.download(m_source, ref.id, ref.version, info.package_dir, m_sink);
+                const auto& action = *actions[i];
+                const auto info = BinaryPackageReadInfo{action};
+                const auto ref = make_feedref(info, "");
 
-                if (res)
+                Path temp_dir = m_buildtrees / fmt::format("upkg_download_{}", info.package_abi);
+                Path zip_path = temp_dir / fmt::format("{}.zip", ref.id);
+
+                const auto result = m_azure_tool.download(m_source, ref.id, ref.version, temp_dir, m_sink);
+                if (result.has_value() && m_fs.exists(zip_path, IgnoreErrors{}))
                 {
-                    out_status[i] = RestoreResult::restored;
+                    out_zips[i].emplace(std::move(zip_path), RemoveWhen::always);
                 }
                 else
                 {
-                    out_status[i] = RestoreResult::unavailable;
+                    msg::println_warning(result.error());
+                }
+                // RemoveWhen mechanism is insufficient since there is now an extra temporary directory 
+                if (m_fs.exists(temp_dir, IgnoreErrors{}))
+                {
+                    m_fs.remove(temp_dir, VCPKG_LINE_INFO);
```

Comment:
Also if this is a directory it probably needs `remove_all`. Also I don't believe you need to check exists first ....